horovod/common/operations.cc

```diff
@@ -190,6 +196,11 @@ OperationManager* CreateOperationManager(HorovodGlobalState& state) {
       new NCCLAllgather(&nccl_context, &gpu_context, &state)));
 #endif
 
+#if HAVE_NCCL && HOROVOD_GPU_ALLTOALL == 'N'
```

Comment:
An interesting case to consider: if the user sets `HOROVOD_GPU_OPERATIONS=NCCL`, then they get `HOROVOD_GPU_ALLTOALL=NCCL` for free.  But if their NCCL version < 2.7, this creates a runtime error.

Alternatively, we could fall back to using MPI All-to-All instead of raising the error, but then users might get poor performance without knowing why (particularly if they specified `HOROVOD_GPU_ALLTOALL=NCCL` explicitly).

I think that overall the way you have it implemented is the best solution, since there is a workaround for users who cannot upgrade NCCL for whatever reason: instead of using `HOROVOD_GPU_OPERATIONS`, set `HOROVOD_GPU_ALLREDUCE` and omit `HOROVOD_GPU_ALLTOALL`.

Maybe we can add some wording to that effect in the exception message when the NCCL version is too low for GPU All-to-All?

Reply:
This aligns with my thinking as well. I thought about having it fallback to MPI but I'd rather have it explicitly break and inform users so they know to update NCCL, rather than silently operating with less performance than expected. 

I beefed up the error message here:
https://github.com/romerojosh/horovod/blob/08116280bb26d9a74bc33f28e90ec47fdece177b/horovod/common/ops/nccl_operations.cc#L589-L593